### ISSUE TITLE
cloudflare_access_application.app_launcher_visible cannot be set to false

### ISSUE THREAD

**Issue by clmaddox** [cloudflare/terraform-provider-cloudflare#1364]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

Terraform v1.1.0
on darwin_arm64
+ provider registry.terraform.io/cloudflare/cloudflare v3.6.0

### Affected resource(s)

cloudflare_access_application.app_launcher_visible

### Terraform configuration files

```hcl
resource "cloudflare_access_application" "example_app" {
  zone_id              = var.cloudflare_zone_id 
  name                 = "example-app"
  type                 = "self_hosted"
  domain               = "example-app.<domain>.com"
  session_duration     = "2h"
  app_launcher_visible = false
}

resource "cloudflare_access_policy" "example_app_policy" {
  application_id = cloudflare_access_application.example_app.id
  zone_id        = var.cloudflare_zone_id 
  name           = "example-app-policy"
  precedence     = "1"
  decision       = "allow"

  include {
    group = [<groupid>]
  }
}
```

### Debug output

[cf_tf_log.txt](https://github.com/cloudflare/terraform-provider-cloudflare/files/7856566/cf_tf_log.txt)

### Expected output

  \# cloudflare_access_application.example_app will be created
  + resource "cloudflare_access_application" "example_app" {
      + account_id                 = (known after apply)
      + app_launcher_visible       = false
      + aud                        = (known after apply)
      + auto_redirect_to_identity  = false
      + domain                     = "example-app.\<domain\>.com"
      + enable_binding_cookie      = false
      + http_only_cookie_attribute = false
      + id                         = (known after apply)
      + name                       = "example-app"
      + session_duration           = "2h"
      + skip_interstitial          = false
      + type                       = "self_hosted"
      + zone_id                    = "\<zoneid\>"
    }

  \# cloudflare_access_policy.example_app_policy will be created
  + resource "cloudflare_access_policy" "example_app_policy" {
      + account_id     = (known after apply)
      + application_id = (known after apply)
      + decision       = "allow"
      + id             = (known after apply)
      + name           = "example-app-policy"
      + precedence     = 1
      + zone_id        = "\<zoneid\>"

      + include {
          + group = [
              + "\<groupid\>",
            ]
        }
    }

### Actual output

  \# cloudflare_access_application.example_app will be created
  + resource "cloudflare_access_application" "example_app" {
      + account_id                 = (known after apply)
      + app_launcher_visible       = false
      + aud                        = (known after apply)
      + auto_redirect_to_identity  = false
      + domain                     = "example-app.\<domain\>.com"
      + enable_binding_cookie      = false
      + http_only_cookie_attribute = false
      + id                         = (known after apply)
      + name                       = "example-app"
      + session_duration           = "2h"
      + skip_interstitial          = false
      + type                       = "self_hosted"
      + zone_id                    = "\<zoneid\>"
    }

  \# cloudflare_access_policy.example_app_policy will be created
  + resource "cloudflare_access_policy" "example_app_policy" {
      + account_id     = (known after apply)
      + application_id = (known after apply)
      + decision       = "allow"
      + id             = (known after apply)
      + name           = "example-app-policy"
      + precedence     = 1
      + zone_id        = "\<zoneid\>"

      + include {
          + group = [
              + "\<groupid\>",
            ]
        }
    }

### Steps to reproduce

1. Add a cloudflare_access_application resource and set app_launcher_visible to false
2. Create a plan with terraform (it will show properly as app_launcher_visible set to false)
3. Apply plan
4. View the application in cloudflare dashboard
5. Observe the app_launcher_visible switch is enabled

### Additional factoids

I tried to track down where the issue may be. There are three pieces (best I can tell) that are involved. Here are the findings for each:
1. Cloudflare terraform provider - The terraform provider seems to be properly setting the flag based on output.
2. Cloudflare Go API client - The application object is passed to this wrapper. Code appears to properly handle the flag, but I suspect the issue is here or in the interface between the provider and the go client.
3. Cloudflare API - I manually sent a curl request with the value set to false and it was created properly.

#### Initial Creation
```
  # module.vpc.module.server.cloudflare_access_application.name will be created
  + resource "cloudflare_access_application" "name" {
      + account_id                 = (known after apply)
      + app_launcher_visible       = false
      + aud                        = (known after apply)
      + auto_redirect_to_identity  = false
      + domain                     = <domain>
      + enable_binding_cookie      = false
      + http_only_cookie_attribute = false
      + id                         = (known after apply)
      + name                       = <name>
      + session_duration           = "2h"
      + skip_interstitial          = false
      + type                       = "self_hosted"
      + zone_id                    = <zone_id>
    }
```
#### Creation log from Access dashboard
```
{
    "domain": <domain>,
    "name": <name>,
    "allowed_idps": null,
    "auto_redirect_to_identity": false,
    "session_duration": "2h",
    "ssh_logging_enabled": false,
    "zoneId": <zone_id>,
    "app_launcher_visible": true,
    "type": "ssh"
}
```
#### Any time a new plan is run
```
  # module.vpc.module.server.cloudflare_access_application.name will be updated in-place
  ~ resource "cloudflare_access_application" "name" {
      ~ app_launcher_visible       = true -> false
        id                         = <id>
        name                       = <name>
        # (10 unchanged attributes hidden)
    }
```
#### Update log from Access dashboard
Old value
```
{
    "domain": <domain>,
    "name": <name>,
    "allowed_idps": null,
    "auto_redirect_to_identity": false,
    "session_duration": "2h",
    "ssh_logging_enabled": false,
    "app_launcher_visible": true,
    "type": "self_hosted"
}
```
New value
```
{
    "domain": <domain>,
    "name": <name>,
    "allowed_idps": null,
    "auto_redirect_to_identity": false,
    "session_duration": "2h",
    "ssh_logging_enabled": false,
    "app_launcher_visible": true,
    "type": "self_hosted"
}
```

**Comment by jacobbednarz** [comment]
please provide the untruncated TF_LOG=DEBUG output as requested. it will contain some context as to what is happening on the API side.

**Comment by clmaddox** [comment]
Updated

**Comment by lwimmer** [comment]
I can confirm this issue with the latest provider (`v3.8.0`) and Terraform (`v1.1.5`).

`app_launcher_visible` cannot be set to `false`, as it is omitted from the PUTed JSON while trying to create or update the application.

It seems that the issue actually lies in the library used by the provider (`github.com/cloudflare/cloudflare-go v0.31.0`), as it defines the attribute `app_launcher_visible` as `bool`, but does set `omitempty`:
https://github.com/cloudflare/cloudflare-go/blob/v0.31.0/access_application.go#L47

The Go JSON marshaller will omit a `false` bool from the JSON if `omitempty` is set.
See also https://github.com/golang/go/issues/13284

**Comment by dani-CO-CN** [comment]
The scope of this bug is larger than reported in this issue, the following flags are all affected (https://github.com/cloudflare/cloudflare-go/blob/v0.35.1/access_application.go#L45-L49):
```
SkipInterstitial
AppLauncherVisible
EnableBindingCookie
HttpOnlyCookieAttribute
```

You can not set these flags to `false` (explicitly) and it depends on the server side default value whether you can set the flag. When the server side default value is `false`, you can set the value to `true`, but only once and can never go back to `false` since the false gets omitted as mentioned above. This bug may even affect all flags in this provider e.g. `https://github.com/cloudflare/cloudflare-go/blob/v0.35.1/access_policy.go#L31`

**Comment by dani-CO-CN** [comment]
I just wanted to add that this also happens with provider `v3.11.0` and terraform `1.1.7`. Another flag that is very likely also effected by this bug is the `http_only_cookie_attribute`, this flag can also never be set to `false` with terraform.  

Interesting side note: the flag `enable_binding_cookie`, which is also bool can be set to `false` with the provider.
